api/envoy/config/listener/v3/listener.proto

```diff
@@ -107,8 +107,15 @@ message Listener {
   // that is governed by the bind rules of the OS. E.g., multiple listeners can listen on port 0 on
   // Linux as the actual port will be allocated by the OS.
   // Required unless *api_listener* or *listener_specifier* is populated.
+  // TODO (soulxu): deprecated the `address` field after `addresses` implemented.
```

Comment:
I didn't mark the old field as `deprecated` now since it will required to modify a lot of example config and test to not use the depcreated field, that will increase this PR size again and making the review harder. But let me know if we have better procedure for this case.

Reply:
I suggest splitting this PR, to a few PRs.
You can start by adding the `addresses` field, mark in not-implemented-hide, and gradually add additional parts of the implementation.
At the end, there should be a well-defined behavior of what happens when both fields (`address` and `addresses`) are provided.

Reply:
Thanks! got it. I will mark it as `not-implemented-hide`, and separate the API change out.

For other parts, also want to hear others' opinions if they are happy to separate the PR, since before people may want to see the full picture. I'm ok with both ways, if we want to separate the whole thing, I will work on it.

@mattklein123 @ggreenway

Reply:
Given that literally every envoy configuration uses these fields, we'll want to support both the old and the new forever. And I'd adopt the semantic that it's a config load error if both of them are set. So it logically turns into a `oneof` between the two, but without modifying the old proto field.

As far as separating the PR or doing it all at once, I'm fine either way. I like seeing how the config will all be put together in one shot, but if the PR gets too big for that I'm fine splitting it. It may be worth defining the entire proto, but just not writing all the code yet, which should be fine if it's `not-implemented-hide`.

Reply:
> Given that literally every envoy configuration uses these fields, we'll want to support both the old and the new forever. And I'd adopt the semantic that it's a config load error if both of them are set. So it logically turns into a `oneof` between the two, but without modifying the old proto field.
> 
I tend to agree with this, but from previous experience, when a configuration server supports both "old" and "new" Envoy clients, and sends them the same config, then the "new" clients will reject it.
One thing we can do is have a runtime guard against the rejection of configs that include both fields, and eventually remove that check.
If `oneof` is the desired goal, consider adding [`oneof_promotion` tag](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md#field-level).

> As far as separating the PR or doing it all at once, I'm fine either way. I like seeing how the config will all be put together in one shot, but if the PR gets too big for that I'm fine splitting it. It may be worth defining the entire proto, but just not writing all the code yet, which should be fine if it's `not-implemented-hide`.

If possible, I think that it makes sense to break this to multiple smaller PRs (not only API vs non-API) to make it more reviewable. That said, if the PR includes mostly new tests, then it should be ok.

Reply:
> I tend to agree with this, but from previous experience, when a configuration server supports both "old" and "new" Envoy clients, and sends them the same config, then the "new" clients will reject it. One thing we can do is have a runtime guard against the rejection of configs that include both fields, and eventually remove that check. If `oneof` is the desired goal, consider adding [`oneof_promotion` tag](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md#field-level).
> 

I also have thing about keeping the old `address` field and adding a new one called `additional_addresses`. I have too much difference with `one_of` way. Only one thing is if the user doesn't specify the `stat_prefix`, then we fallback to use `address` field as stat_prefix. For the `oneof` case, if only the `addresses` specified, then choose the first one as the stat_prefix. The `additional_addresses` way has a little better sematic here.

> > As far as separating the PR or doing it all at once, I'm fine either way. I like seeing how the config will all be put together in one shot, but if the PR gets too big for that I'm fine splitting it. It may be worth defining the entire proto, but just not writing all the code yet, which should be fine if it's `not-implemented-hide`.
> 
> If possible, I think that it makes sense to break this to multiple smaller PRs (not only API vs non-API) to make it more reviewable. That said, if the PR includes mostly new tests, then it should be ok.

Let me see what the separate PR looks like. I found I can't change `ListenerImpl` and `ListenerManagerImpl` separately, since our unit-tests test them together. And we can live this PR here for people who want to see the whole picture.

Reply:
> Let me see what the separate PR looks like. I found I can't change `ListenerImpl` and `ListenerManagerImpl` separately, since our unit-tests test them together. And we can live this PR here for people who want to see the whole picture.

Just to clarify, breaking to multiple PRs is not a must, and in some cases it may make sense to have everything in one PR.

Reply:
I should be able to separate the PR as commits have in the PR https://github.com/envoyproxy/envoy/pull/19367/commits, I checked each of them has unittest covered. The last PR should be for the integration tests.